File: cplusplus/src/solutions/Solution0005.cpp
#include "Solution0005.hpp"

#include <algorithm>
#include <iostream>
#include <vector>

Solution0005::Solution0005()
    : ISolution{}
{
}

bool Solution0005::isReady() const
{
    return true;
}

void Solution0005::describeIssue() const
{
    logWarn() << "Longest Palindromic Substring";
}

void Solution0005::run()
{
    std::string checkingString{"123bbbbbbbbbbbbaaaaaaaaabbbbbbbbbbbbbbbbbbb321"};

    logVerbose() << "Input params:";
    logVerbose() << "   1) string : " << checkingString;

    const auto palindrome = longestPalindrome(checkingString);
    logWarn() << "The palindrome is   : " << palindrome;
}

std::string Solution0005::longestPalindrome(const std::string& s)
{
    const int size = s.size();
    std::string palindrome{s[0]};
    int startPalindromeIndex = 0;
    char startPalindromeChar = s[startPalindromeIndex];
    int endPalindromeIndex = size - 1;

    while (startPalindromeIndex < size - palindrome.size()) {
        // find equal letters
        while (startPalindromeChar != s[endPalindromeIndex]) {
            --endPalindromeIndex;
        }

        const auto newPalindromeSize = endPalindromeIndex - startPalindromeIndex + 1;
        if (palindrome.size() < newPalindromeSize) {
            int leftLetterIndex = startPalindromeIndex + 1;
            int rightLetterIndex = endPalindromeIndex - 1;
            while (leftLetterIndex < rightLetterIndex) {
                if (s[leftLetterIndex] != s[rightLetterIndex]) {
                    break;
                }
                ++leftLetterIndex;
                --rightLetterIndex;
            }
            if (leftLetterIndex >= rightLetterIndex) {
                palindrome = s.substr(startPalindromeIndex, newPalindromeSize);
                ++startPalindromeIndex;
                startPalindromeChar = s[startPalindromeIndex];
                endPalindromeIndex = size - 1;
                continue; // while for startPalindromeIndex
            }
        }

        endPalindromeIndex -= 1;
        if ((startPalindromeIndex >= endPalindromeIndex)
            || (palindrome.size() > endPalindromeIndex - startPalindromeIndex + 1)) {
            ++startPalindromeIndex;
            startPalindromeChar = s[startPalindromeIndex];
            endPalindromeIndex = size - 1;
        }
    }

    return palindrome;
}

File: python/solutions/Solution0005.py
"""
0005 Longest Palindromic Substring
Given a string s, return the longest palindromic substring in s.

Example 1:
    Input: s = "babad"
    Output: "bab"
    Explanation: "aba" is also a valid answer.

Example 2:
    Input: s = "cbbd"
    Output: "bb"

Constraints:
1 <= s.length <= 1000
s consist of only digits and English letters.
"""

from solutions.ISolution import ISolution

class Solution0005(ISolution):
    def isReady(self):
        return True
    
    def describeIssue(self):
        print("[Solution 0005]: Longest Palindromic Substring")

    def run(self):
        checkingString = "123bbbbbbbbbbbbaaaaaaaaabbbbbbbbbbbbbbbbbbb321"
        print(f'Input params:')
        print(f'    1) string: {checkingString}')
        palindrome = self.longestPalindrome(checkingString)
        print(f'The palindrome is: {palindrome}')

    def longestPalindrome(self, s: str) -> str:
        S_LENGTH = len(s)
        palindrome = s[0]

        startPalindromeIndex = 0
        endPalindromeIndex = S_LENGTH - 1
        while startPalindromeIndex < S_LENGTH - len(palindrome):
            # find equal letters
            while s[startPalindromeIndex] != s[endPalindromeIndex]:
                endPalindromeIndex -= 1

            newPalindromeSize = endPalindromeIndex - startPalindromeIndex + 1
            if startPalindromeIndex >= endPalindromeIndex or len(palindrome) >= newPalindromeSize:
                startPalindromeIndex += 1
                endPalindromeIndex = S_LENGTH - 1
            else:
                newPalindrome = s[startPalindromeIndex:endPalindromeIndex+1]
                if newPalindrome == newPalindrome[::-1] :
                    palindrome = newPalindrome
                    startPalindromeIndex += 1
                    endPalindromeIndex = S_LENGTH - 1
                else:
                    endPalindromeIndex -= 1

        return palindrome